### ISSUE TITLE
Ignore changes for a single AWS Resource Tag

### ISSUE THREAD

**Issue by emoshaya** [hashicorp/terraform#6632]
Is there a way to limit the @ignore_changes to just a single tag e.g. VERSION tag. As apposed to every single tag for a particular resource? I have a VERSION tag which is updated regularly by Jenkins and I would like a way to just ignore_changes for that tag.

Thanks

**Comment by dsolsona** [comment]
I'm also interested in this. `ignore_changes = ["tags"]` seems very generic and it appears that it also stops terraform adding new tags to resources.

**Comment by jen20** [comment]
Hi @emoshaya-cognito! There currently isn't a way to ignore changes to individual map fields. I don't believe there is a way to implement this given the current semantics of ignore_changes, though @phinze is more familiar with that area of the code base than I am.

**Comment by manojlds** [comment]
Would also like to have this. Use case - EBS volume tags are updated with the instance id of the instance in which they are mounted. Would like to control other tags from terraform, but ignore changes to the instance id tag.

**Comment by jimhoagland** [comment]
I would also like this.  We run JanitorMonkey regularly, which adds CreatorId, CreatorName, PrincipalId.  I really don't want those removed when doing an apply.  Perhaps those can be listed in the config with some kind of marking that those tags should be ignored if present.

**Comment by pgporada** [comment]
This would be useful for us as well. Amongst other things we're doing is setting the current version of our application in the AWS console so people in there can check it without too much effort.

**Comment by Ashald** [comment]
We would be glad to see this as well - we need to set a date-derived tags upon resource creation but the change in such tag shouldn't be a reason for resource "invalidation".

**Comment by nwalke** [comment]
We have an application that creates/updates/deletes tags on the resources it's running on, but have a set of tags that never get removed or updated.  Would be super nice to ensure that group of tags never gets touched, but disregard the rest.

**Comment by joseph-holland** [comment]
Would also like this functionality.

**Comment by MikeMcMahon** [comment]
We do this already and it worked flawlessly in older versions of terraform (0.8.8 and prior)

```
ignore_changes {
  "tags.%",
  "tags.hostname",  # dynamically generated on boot time
   ...
}
```

However it seems that now it still seemingly ignores the individual tags (as it always has) but now for some reason in TF 0.9.11 it tries to re-apply the tags over and over and over and over and only seems to actually ignore the tag in the apply, not in the plan.

**Comment by apparentlymart** [comment]
Hi all! Thanks for the great discussion here.

This is hard to implement today in Terraform core due to how Terraform currently thinks of map attributes. Indeed, @MikeMcMahon's workaround above is depending on an implementation detail of how maps are represented internally that is very likely to break in future if it hasn't broken already, since Terraform isn't really "understanding" what's going on there and instead it is (or was) just working by accident.

We are starting to plan some changes to the configuration language that would shore up how Terraform thinks about maps internally. While adjustments to the behavior of `ignore_changes` are unlikely to be included in the first pass (that goes deeper into Terraform's guts than we're currently looking at) we will hopefully be able to get some better building blocks in place to support this in a smarter way in future. Unfortunately this is one of those things where there's a fair amount of supporting work that needs to be done first.

In principle we could do something special in the AWS provider to deal with this use case specifically, rather than changing Terraform core, but given how many different resources deal with `tags` it'd take some significant changes across many resources to get this implemented in a robust way even just within the provider itself.

So with all of that said, this _is_ something we want to support, but it's unfortunately still some ways out. :confounded:

**Comment by mr-olson** [comment]
From the comments so far, it seems there are 2 use cases (both of which would be useful for my needs)

1. ignore_changes per tag - original report

2. Ignore unknown tags automatically - create & manage *only* the set of tags specified in the template - if tags Foo and Bar are set in TF, but Fiz and Baz are later set via some other process, TF should only consider Foo and Bar's presence & values when calculating changes (@nwalke's comment & others).

**Comment by flypenguin** [comment]
yup, also hit that use case right now. any updates on this?

mine is kops deployments in AWS - it adds tags to terraform-created resources (namely subnets), and TF wants to remove them later.

**Comment by jimbeck** [comment]
@flypenguin we just hit the same issue with using kops to create a k8s cluster. Did you do anything to get around this? Minus handjamming those tags into the subnet creation >.<

**Comment by flypenguin** [comment]
nah, I didn't. Actually it *seems* that the removal of the tags does not harm the cluster or kops, at least not in my initial tests. The only thing you get is a tagging battle between kops and terraform, which is something I can live with (if there are really no side effects) until HCL and terraform are getting straightened out a bit. I guess everybody is using HCL in ways nobody thought about in the beginning, and it is hitting its limits :)

**Comment by sidick** [comment]
I'm actually seeing the same issue when creating an EKS cluster which adds tags to the VPC it's associated with

**Comment by flypenguin** [comment]
Well, it *does* affect kops clusters. If you deploy an `nginx-ingress`, it will try to use the subnet tags to find suitable subnets for the ELB. No tags, no ELB.

This is something I *really* would like to see now.

**Comment by jeremygaither** [comment]
Any chance this might be coming with Terraform 0.12? This would be very helpful with multiple scenarios, especially with kops-generated Terraform and general version tags.

**Comment by apparentlymart** [comment]
We've not intentionally made any changes in this area for v0.12, but it's possible that some of the changes we've made for other reasons might come together to offer at least a partial solution here. Once the v0.12 branch lands in master (still quite a bit of work to do there first, unfortunately) we'll give this a try and see what the situation is.

The foundational piece that has shifted for the v0.12 release is that Terraform will now parse `ignore_changes` as an expression relative to the resource itself, so _syntactically_ it's in theory possible to state the following:

```hcl
  ignore_changes = tags["Version"]
```

...however, there's of course more to `ignore_changes` than just syntax, and the v0.12 scope is already pretty large, so if this isn't fully solved by the other changes then we'll probably need to hold on this one for a subsequent release, but at least the building blocks will be in place to make this more reasonable to implement.

I've relabeled this issue so it'll be on our radar to review it once the development branch is merged and we're testing prior to the release.

**Comment by richievos** [comment]
@apparentlymart would you expect those changes to support wildcarding? For the kops example that some have mentioned, the tag name would be something of the form ```kubernetes.io/cluster/foo-bar-123```, so the ignore would likely need to be ```ignore_changes = tags["kubernetes.io/cluster/*"]```.

**Comment by lopezm1** [comment]
excited for this functionality

**Comment by dkelertasaamgroup** [comment]
waiting for this too - need to ignore unmanaged tags

**Comment by galdor** [comment]
We are using the following: 

```
  lifecycle {
    ignore_changes = ["tags.ServiceVersion"]
  }
```

but still have the same issue than MikeMcMahon above, with Terraform trying to apply:

```
tags.%:                                "5" => "4"
```

For all instances.

Anything new on this feature ?

**Comment by andyspiers** [comment]
@galdor if you're trying to stop terraform removing a tag that was added outside of terraform then you need to ignore both the tag(s) that were added and the **number** of tags:
```hcl
lifecycle {
    ignore_changes = ["tags.ServiceVersion", "tags.%"]
  }
```
I tested this just now on a DynamoDB table resource but it should be the same for all resource types.

(Note that ignoring `"tags.%"` is not needed if terraform expects the tag to be there but you just need to ignore changes to the value)

**Comment by galdor** [comment]
It works indeed, thank you !

**Comment by enekofb** [comment]
In the context of kubernetes the following works ...

```
resource "aws_subnet" "xxxx" {
  ...
  lifecycle {
    ignore_changes = ["tags.%","tags.kubernetes"]
  }
}
```
tested with 

```
Terraform v0.11.10
+ provider.aws v1.43.2
```

**Comment by szinck** [comment]
@enekofb worked for me! Thanks! If you are using kops you must also ignore tags.SubnetType.

**Comment by citizenkahn** [comment]
hmmm.  So I'm using a module and creating an ASG.  I've try many things in tf 0.11.7 and 0.11.10 and only ignoring all tags works:

`Ignores`:

```
lifecycle {
    ignore_changes = ["tags.CreatedAt", "tags.%"]
  }
```

`Does not Ignore`:
```
lifecycle {
    ignore_changes = ["tags.CreatedAt", "tags.%"]
  }
```

Providers
- "null" (1.0.0)...
- "tls" (1.2.0)...
- "aws" (1.52.0)...
- "template" (1.0.0)...
- "local" (1.1.0)...

I'm not sure what's wrong.

**Comment by DrmCtchr** [comment]
in terraform i have 
  tags = {
    tag1 = "Hello"
   tag2 = "world"
  }

outside of terraform i have my resource additionally tagged with tag3 = "foo" and tag4 = "bar"

I need terraform to apply tag1 and tag2 but ignore any changes to tag3 and tag4 and any additional tags that are dynamically applied by tagging engines outside of terraform that are unknown at the time of terraform resource creation.

I have tried 

lifecycle {
    ignore_changes = ["tags"]
  }
 
but that will cause terraform to **not** apply tag1 and tag2 on initial resource creation.

**Comment by bpoland** [comment]
@DrmCtchr several folks above have mentioned ignoring the specific tag names plus the number of tags, like this:

```
lifecycle {
  ignore_changes = ["tags.%","tags.tag3","tags.tag4"]
}
```

That has worked perfectly for me.

**Comment by DrmCtchr** [comment]
right the problem is when the tag names are not known ... since our organization tags our resources for us the only thing we can be sure of is the names of tag1 and tag2 after that all bets are off. We need to apply tag1 and tag2 but we cannot wipe out tag3 and tag4 and tag239842387 etc. on reapply

**Comment by bpoland** [comment]
Ah okay.  Yeah that's tough then.  I would have thought the way you have it above would have created tag1 and tag2 on initial creation, and then not touched the tags after that.

**Comment by jbscare** [comment]
This works for us:
```
lifecycle {
ignore_changes = ["tags.%", "flexy"]
}
```
where `flexy` is a tag that that Terraform creates at provisioning time with a default value, but then we want to change it by hand later, and don't want Terraform to complain about it.

Ignoring `tags.%` causes it to ignore any other tags that it didn't create. It continues to manage all the other tags that it created.

@DrmCtchr , in your last comment, you didn't mention that you were ignoring `tags.%` -- are you? If not, try ignoring just that, and see if it does what you want.

**Comment by DrmCtchr** [comment]
In testing this:
lifecycle {
ignore_changes = ["tags.%"]
}

DID NOT WORK It did ignore the number of tags ... but in the plan the rest of the tag values set out of band were modified to ""

**Comment by jbscare** [comment]
> in the plan the rest of the tag values set out of band were modified to ""

Just to clarify, you mean additional tags that were added that Terraform doesn't know about? Or tags that Terraform *does* know about, but that you modified out of band?

If you want Terraform to create a tag and set an initial value, and then ignore changes to it after that, you need to add it to the ignore list, like 
```
ignore_changes = ["tags.%", "flexy"]
```
If Terraform is blanking the values of tags that it doesn't know about, even with `tags.%` being ignored, that's surprising to me.

**Comment by DrmCtchr** [comment]
right if i have terraform set "tag1" to 'foo' and "tag2" to 'bar'
and have some non-terraform automation comes around and  sets "tag3" to 'narf' and 'tag4' to "troz"

then even if : ignore_changes = ["tags.%"]

then  on next plan: 
tag3 = "narf" -> "" and tag4 = "troz" -> ""

**Comment by DrmCtchr** [comment]
the BUG is that if you have:

tags {
"tag1" = "foo"
"tag2" = "bar"
}

AND you have:
ignore_changes = ["tags"]

that the ignore gets evaluated *after* the tags block effectivey ignoring it... so you will never get tag1  or tag2 applied.

**Comment by lopezm1** [comment]
Back on this thread after 8 months.

To clarify for everyone... 

Most people on this thread are trying to ignore a single tag when created with the `tag` resource on the ASG resource. NOT the `tags` resource. This is an important difference. We are using `tag` because we need to propagate the tag on launch. 

We are using...
```
resource "aws_autoscaling_group" "asg_generic" {
  .
  .
  .
  tag {
    key                 = "ServiceName"
    propagate_at_launch = true
    value               = "${var.asg_tags["Version"]}"
  }

  lifecycle {
    ignore_changes = ["tag", "min_size", "max_size", "desired_capacity"]
  }
}
```

Not....

```
tags {
    Version = "HelloWorld"
}
```

Therefore, `ignore_changes = ["tags.%", "Version"]` or `ignore_changes = ["tags.%", "tags.Version"]` does not really apply to us. 

**Most people are looking for ** `ignore_changes = ["tag.Version"]` **so we can ignore a single tag used that is propagating that tag on launch**


See everyone in another 8 months. Hopefully we have v0.12 by then 😆

**Comment by pgporada** [comment]
The following statement was broken after upgrading from `0.11.14` => `0.12.0`. I think my `tags.%` would prevent any `tags` from being changed, but that's not of concern to me right now.
```
  lifecycle {
    create_before_destroy = true
    ignore_changes        = ["tags.%", "tags.kubernetes.io"]
  }
```

Here's the `terraform validate output
```
$ terraform validate

Error: Attribute name required

  on modules/vpc/vpc.tf line 28, in resource "aws_vpc" "ct":
  28:     ignore_changes        = ["tags.%", "tags.kubernetes.io"]

Dot must be followed by attribute name.


Error: Attribute name required

  on modules/subnets/template/subnet_tmpl.tf line 55, in resource "aws_subnet" "template":
  55:     ignore_changes        = ["tags.%", "tags.kubernetes.io"]

Dot must be followed by attribute name.


Error: Attribute name required

  on modules/subnets/template/subnet_tmpl.tf line 55, in resource "aws_subnet" "template":
  55:     ignore_changes        = ["tags.%", "tags.kubernetes.io"]

Dot must be followed by attribute name.
```

**Comment by apparentlymart** [comment]
Hi @pgporada,

Please open a new issue for that, including the information requested in the issue template. If your configuration still contained a `tags.%` entry after running through the upgrade guide then that's a bug in the upgrade tool, which is something different than what this issue is about.

**Comment by cmbrad** [comment]
Has anyone found a way to do ignore a single tag successfully in 0.12? I tried:

ignore_changes = [tags.Name]

Accepted by the parser but seemingly does nothing. Plans still try and modify the Name tag

ignore_lifecycle = [tags["Name"]]

Generates the error "A static variable reference is required."

ignore_lifecycle = [tags]

This works as expected but ignores all tags.

**Comment by kladiv** [comment]
+1

**Comment by akdracarys** [comment]
+1 to ignore a single tag in 0.12

**Comment by davi5e** [comment]
Read it all and my problem is essentially the same as everyone described, expect that I'm not creating the resources myself, I'm using modules...

As in [terraform-aws-modules/vpc/aws](https://registry.terraform.io/modules/terraform-aws-modules/vpc/aws/2.7.0) for VPC and then [terraform-aws-modules/eks/aws](https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/5.0.0) for EKS.

As reporte, the Kubernetes tags are `"kubernetes.io/cluster/*" = "shared"` and they're added to each subnet and also for the VPC resource.

I'm using 0.12 but, even if `lifecycle.ignore_changes` were to work, how would it be used **from the module**? =/
